### DISCUSSION TITLE
Purpose of qpdf and target audiences

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#1] [Mission]
The qpdf readme file states

> qpdf is a command-line tool and C++ library that performs content-preserving transformations on PDF files. It supports linearization, encryption, and numerous other features. It can also be used for splitting and merging files, creating PDF files (but you have to supply all the content yourself), and inspecting files for study or analysis. 

This can be  broken down into 4 parts

1. qpdf is a command-line tool and C++ library that performs content-preserving transformations on PDF files. It supports linearization, encryption, and numerous other features. 
2. It can also be used for splitting and merging files, 
3. creating PDF files (but you have to supply all the content yourself)
4. It can also be used for inspecting files for study or analysis. 

This raises immediately two questions

1. is this list complete?
2. are they in priority order? (My view would be that 1 - 3 are of equal priority and 4 is of lesser priority).

The above suggests to me that there are 3 or possibly 4 target audiences that we are trying to cater for

1. end users of the qpdf tool and of applications developed with the library
2. developers using the library
3. users that want inspect PDF files. This includes 
  a.  end users and developers inspecting files encountered as end users or developers, and
  b. users wanting to inspect more general files, including severely damaged files.
5. possibly users that want to create PDF files for study and experimentation

Are there any other users that we need to consider?
Are some users more key than others? (My view would be that 1 followed by 2 are key, closely followed by 3a. 3b and 4 are of less importance.)

The interests of what I consider the key groups of users include

#### 1. end users

  - reliability
    - produce correct outputs
    - when failing, fail in an obvious way
  - functionality
    - support all the transformations/operations the user requires
  - scope
    - work with the widest possible range of inputs
      - size of files
      - damaged files
    - allow files to be used by the widest range of readers
      - produce conformant PDF files 
  - information
    - have sufficient information to be alert to possible problems with inputs and/or outputs
    - don't have unnecessary information that may obscure the necessary info 
  - performance

#### 2. developers
  - most/all of the end user requirements
  - easy to use API
  - prevent / detect programmer error
  - sufficient flexibility to work around limitations of qpdf, i.e. have low level access where needed
  - additional information to assist debugging

#### 3a.  end users and developers inspecting files
  - have sufficient information to identify and analyse problems in potentially viable files
  - identify severely damaged files

If the above capture the key aims, users and interests of those users, the two most obvious questions are:

1 Should we slow adding new functionality by an attempt to capture information that is not required for for the above?

2 Are we providing the right information to the right users? For examples, 

- When errors are detected, are most end-users interested in objgens and offsets where the issue occurred?
  Should this information be suppressed  by default, except for debug builds? For the CLI we could have a message like
  "Succeeded with warnings .... . For further detail, rerun with --verbose flag". Providing the information unnecessarily has a significant   impact on the memory footprint and some impact on performance.

- Has the message " -> null returned from getting key $VD from non-Dictionary" any use other than during debugging? So again, should such messages be suppressed  by default except for debug builds?

**Comment by jberkenbilt** [comment]
Any reason not to create a mission category in qpdf and discuss this there out in the open? Then more people will see it and be able to share their views on how they think qpdf should be used.

**Comment by jberkenbilt** [comment]
It's really hard for me track with the above list, so let me ramble a little and share my thoughts.

When I created qpdf in 2002 and made it open-source in 2008, the PDF spec was a bit of a moving target, and there were very few open-source tools that provided low-level access to PDF, and as far as I know, none in C/C++. At that time, my main languages were C++ and perl. qpdf was the first open source tool I know of that could linearize files or support AES256 encryption, and when it gained the ability to split and merge pages, there weren't other native C/C++ based tools that could do it to my knowledge.

The landscape has changed a lot of the last 20+ years. C/C++ are no longer languages that most people use as their main languages. They still rule when it comes to system-level tools in Linux and are still in wide use in many areas, but people who need to work with PDF files in new code are, in my experience, more likely to be working in JavaScript, C#, Java, Go, Python, or even rust. Some of these languages have decent libraries that allow people to create PDF without really having to understand PDF internals in any way.

qpdf has tried to be different things at different stages of its life. To me, the most important offering of qpdf is the ability to get low-level access to PDF in a safe and performant way. Its API allows you to usually not have to care whether an object is direct or indirect, and it gets out of your way, allowing you to directly manipulate content and structure in PDF files. It has very few dependencies and can be built on a wide variety of systems. To me, these are its biggest strengths. Features like qpdf JSON and QDF, I believe, differentiate qpdf from other tools.

It is possible to use qpdf to create PDF files from scratch, but I would consider this a less important use of the library. That may be good for simple cases, but since qpdf doesn't have awareness of fonts or the ability to support some of the compression filters, its use in this space is limited, and there are better choices.

The API tries to create higher-level abstractions for certain things using a "helper" pattern over a "wrapper" pattern. I like light abstractions that don't try to hide you from underlying details but rather take care of some of the trickier or more error-prone aspects of things. For example, the page helpers allow you to add/remove pages and take care of keeping the pages tree consistent. There are similar abstractions that help you keep track of name and number trees, the linkage between form fields and annotations, the various matrix operations connected with displaying annotations, and other tactical areas within the PDF. Some might criticize these as leaky abstractions, but their intention is not to be opaque. That said, I may have missed the mark a little. In some cases, QPDFObjectHandle exposes way too much stuff, some of which better belongs to helpers, but much of it predates helpers, and I have had a rigorous adherence to backward compatibility. Also, some of the helpers maintain caches. While my intention was to have them created ephemerally, used for a local change, and destroyed, this is fairly inefficient. It is also the case that the helpers have mostly only been implemented far enough to provide support for specific qpdf features and may not be complete.

A significant area of growth for qpdf would be implementing the pages epic. I see this as consistent with the vision because qpdf would abstract away the work required to keep document-level structures consistent with page-level structures. Using qpdf to create things like outlines or links would still be hard, but once they were there, qpdf would be able to keep them updated as pages are shuffled around. The pages epic discusses an API for this, but it's not fully fleshed out, and it doesn't really feel very natural to me. I'm concerned about there being too many layers of abstraction that qpdf can operate with. The qpdf API is already bloated and inconsistent.

qpdf would benefit from supporting more types of compression, certificate validation, full support for incremental update, and other things like that that would enable qpdf to do what it does well with a wider range of files.

In general, I have been more focused on having qpdf preserve compliance with the spec for files that are already compliant than having it stop users from creating files that are not compliant. While it is useful to have qpdf be able to create intentionally non-compliant files for testing purposes, this is a corner case and is not a justification for not adding more checks for spec violation. That said, putting a lot of work into that area feels low-value to me because people who are doing complex from-scratch kinds of operations are probably better off using other libraries.

So...features that enable ways to get into PDF files at a low level and inspect or update them with surgical precision are consistent with my view of qpdf's mission. This would include incremental update support for writing, JSON, and QDF as well as an option to QPDFWriter that would aim to preserve original object IDs. Features that enable qpdf to read more types of files, such as supporting more x509 features or more compression filters would be valuable, but not if it means adding additional dependencies. It would be better to make it easy for people to extend the library. qpdf's custom filters are supposed to do that, but I don't know if anyone has ever used them. Certainly features to make it easier to preserve document-level structures when manipulating pages would be consistent with qpdf's mission.

Things that I would consider out of scope would be anything dealing with font subsetting, text extraction, or content creation. One place where qpdf deviates from this is appearance stream generation, but qpdf doesn't do this all that well or support all the flags, so that may be better as an example of why I didn't want to take qpdf in that direction.

I don't think I can write a clear mission statement, but I can definitely comment on specific ideas and describe why I think they do or don't fit with what qpdf is trying to do. That said, if I hand over some of these decisions to co-maintainers or the community, then others may decide to take qpdf in a different direction.

In summary, I think qpdf should aim to be better at the things that it is most strongly suited for, such as low-level access to PDF constructs and ability to manipulate the pages in PDF files, and I think qpdf should draw a line at font and content management or interactive features, leaving those to other applications and libraries.

I also think it may be time to work toward a pivot where we break what needs to be broken and fix some old mistakes in the API. This could include adding more type safety to QPDFObjectHandle, getting rid of a lot of the cruft in the QPDFObjectHandle API, and removing things in the qpdf API that would be better served in other ways with more modern versions of the C++ standard library.

I hope this input is useful. The truth is that I don't have a clear, succinct picture in my mind of qpdf's mission. What I have is a sense of what I think fits and I think doesn't, but it's more of "I'll know it when I see it" than any hard and fast rule, and I am probably inconsistent about things close to the edge.

**Comment by m-holger** [comment]
> I hope this input is useful. The truth is that I don't have a clear, succinct picture in my mind of qpdf's mission.

It is useful. I am not really after defining a clear mission statement. My aim is to establish enough of a common view on where qpdf should be going to allow us to prioritize work in the short and medium term and make a plan for qpdf 11.10, 11.11 and 12.0.

To pick up on just one point:

> To me, the most important offering of qpdf is the ability to get low-level access to PDF in a safe and performant way. 

I think qpdf is doing pretty well in this area as far as the objects defined in section 7.3 of the spec is concerned. Where I see scope for development is at the next level up that is used in much of the rest of the spec including the tables defining the the various dictionary objects.
i.e. at the level of e.g. optional names or 'name or array of names'. This is level of detail that users should not have to worry about and that can be abstracted away without loosing the ethos of providing low level access.

The other area of low level access that in my view has scope for improvement are operators. At the moment qpdf basically just provides operator names, or rather, it treats (when parsing contents) any name that it does not recognize as operator. While I agree that understanding the semantics of operators is (mostly) out of scope for qpdf, I feel understanding the syntax ought to be in scope - at a minimum determining whether the operator name is valid but preferably determining whether the sequence of operands and operators is valid and packaging operators with their operands. I believe pikepdf already does this and I think it would be worth asking @jbarlow83 whether he would be willing to contribute that code to qpdf.

**Reply by jbarlow83** [reply]
I did write a content stream parser that groups operands with their operator, and packages inline images into a single Python object. It does not try to work out if the sequence is valid. 

I also wrote a content stream builder that lets you construct a content stream in naturally in Python, along these lines, to do low level drawing:

```python
# q 10 0 0 10 0 0 cm ...
canvas.do.push().cm(Matrix(10,0,0,10,0,0)).fill_color(Color(255,0,0).rect(1,1,2,2).pop()
```

But it's only a generator and doesn't implement every feature.

**Reply by jbarlow83** [reply]
You can use the code if you find anything useful there in, but what I did has more to do with Python support than PDF features.

The trouble with any sort of content stream parsing is that there are files with broken content streams out there that PDF viewers seem to shim into displaying correctly. There's also PDFs that exceed the hard limits in the PDF spec, like the graphics stack depth, and still work fine. It's tricky to explain content stream issues, tricky to fix them, and hard to tell users what they should do to fix them.

**Comment by m-holger** [comment]
From #10 

> even repair of the pages tree should be triggered only by doing something with the pages tree. If qpdf doesn't have any reason to traverse the pages tree, then it should not touch it or repair any pages because doing so removes the ability to use qpdf as a tool for inspecting broken files and trying to see what's wrong with them, and that does remain an important use case for qpdf.

This seems to me to be one of the cases where the needs of the different purposes of qpdf are contradictory, and without a conscious effort to clearly distinguish between them there is the danger that none of them will be served optimally.

When viewed as tool for inspecting PDF files it clearly makes sense to minimise repairs and other transformations, and arguably qpdf already interferes too much. Similarly. when viewed as a library that provides very low-level access, repairs ought to be optional.

However, when qpdf is viewed as a tool for transforming, splitting and merging PDF files the aim ought to be to produce files that are as widely usable as possible, which means producing outputs that conform with the standard whenever possible. It also means that the behaviour of qpdf should be consistent - it should not e.g. repair the pages tree sometimes and but not at other times, based on conditions that are only obvious if you have a good understanding of how qpdf is actually implemented.

> I've generally taken a policy of not fixing things that qpdf doesn't have to fix to do its job because I don't want qpdf to make the situation worse for broken files with regard to being viewable in other places. In other words, I don't like getting bug reports that say, "I could open this in Acrobat before I ran it through qpdf, and now I can't."

That is a valid concern. That said, there are certain repairs that appear reasonably safe (e.g. remove non-dictionary nodes from the pages tree or make a direct object indirect to conform with the spec). Where a repair may not necessarily  be safe, in my view the best solution would be to provide a suitable warning and make the repair optional.

From a practical point of view, it is much easier to have confidence in higher level operations if validations and repairs are carried out early. In my view, if qpdf is not used to convert a file to qdf or JSON, or one of the inspection options is used, the trailer, catalog and pages tree should be checked and if necessary repaired before any other operations are carried out.
